### PR TITLE
サンプル数に関する罰則項の導入

### PR DESCRIPTION
**Pull request by ciela**
サンプル数に関して罰則を課す実装をOn-policy実装（キューイングなし）に追加してみました。
configにて `adoption_size_penalty` という比率を調整することで罰則の効き具合を調整できます。

### CODE REVIEW

File: joeynmt/model.py [hunk 1]
```diff
@@ -592,6 +595,11 @@ def adoption_model(log_prob: Tensor, tau: Tensor) -> Tensor:
                     trg = trg.index_select(0, adopted_indexes)
                     length_norms = length_norms.index_select(0, adopted_indexes)
                 # adoption end
+                # compute adoption size grads and backward
+                adoptsize_loss = -adoption_size_penalty * (adopted_size - prev_adopted_size) * log_probs
```
Comment (ciela) [inline joeynmt/model.py:599]: 擬似コードにおける `adopt_size` はサンプル数のスカラー値ということでよろしかったでしょうか？
（下の \| S_{prev} \| との差分を取るときをどう表現して良いのかちょっと分からず・・・）
image

Comment (ciela) [reply joeynmt/model.py:599]: またここでこれまで累積された `log_probs` をモデルによる尤度（擬似コードにおける $\pi_\theta$ ）として用いていますが、これは累積されてないもの（ `torch.log_softmax(logits)` の値）を用いたほうがよろしかったでしょうか？

Comment (tetsuromorimura) [reply joeynmt/model.py:599]: これはバッチ単位の処理ではなく、バッチ内の各サンプル単位での処理になります。今のadopted_sizeはどのサンプル（batch_index）かを考慮せず、バッチ内のトークン列を一括処理したものだと思うので、batch_indexについてループを回すなどして計算する必要があると思います。

累積した方になります。ただlogをとらない単なる$\pi_\theta$なので、PGの部分のロストは若干異なりますー

Comment (ciela) [reply joeynmt/model.py:599]: > これはバッチ単位の処理ではなく、バッチ内の各サンプル単位での処理になります。今のadopted_sizeはどのサンプル（batch_index）かを考慮せず、バッチ内のトークン列を一括処理したものだと思うので、batch_indexについてループを回すなどして計算する必要があると思います。

あーなるほど、各サンプルごとについてサンプルされた数が `adopt_size` で、そこにその時のトークン尤度を足し合わせていって最終的な `adopt_size` としている、ってことですかね。

> 累積した方になります。ただlogをとらない単なる$\pi_\theta$なので、PGの部分のロストは若干異なりますー

こちらも承知しましたー

**Comment by ciela** [comment]
@tetsuromorimura Draftを外しましたー、Huddleで話していた内容について実装を行ったのでご確認お願いしまっすー

File: configs/sbp_in_domain.yaml [hunk 2]
```diff
@@ -58,13 +58,14 @@ training:
       alpha: 1
       reward: "bleu"
       samples: 5
-      max_adoption_size: 100
+      max_adoption_size: 5
```
Comment (tetsuromorimura) [inline configs/sbp_in_domain.yaml:61]: 確認です。この変更はロギングのためですか？
現状、on-policyの場合、`max_adoption_size`は↓のワーニング出力のためにしか使ってないと思いますので。
https://github.com/ciela/reinforce-joey/blob/0c7ca4607815719e6ab8a1131e8d95e6e6ebf4b9/joeynmt/model.py#L580

File: joeynmt/model.py [hunk 3]
```diff
@@ -588,9 +591,15 @@ def adoption_model(log_prob: Tensor, tau: Tensor) -> Tensor:
...
                     src_mask = src_mask.index_select(0, adopted_indexes)
+                    sum_probs = log_probs.exp().sum(dim=1).index_select(0, adopted_indexes)
                     log_probs = log_probs[to_adopt].unsqueeze(dim=1)
                     trg = trg.index_select(0, adopted_indexes)
                     length_norms = length_norms.index_select(0, adopted_indexes)
+                    alive_batches = alive_batches.index_select(0, adopted_indexes)
+                    # compute adoption size penalty
+                    uniqs, counts = alive_batches.unique(return_counts=True)
+                    adopt_size = torch.stack([sum_probs.masked_select(alive_batches == i).sum() for i in uniqs])
+                    (-adoption_size_penalty * torch.pow(adopt_size - counts, 2).sum()).backward(retain_graph=True)
```
Comment (tetsuromorimura) [inline joeynmt/model.py:602]: `(adopt_size-counts)**2` は小さくしたいものなので、`-adoption_size_penalty`のマイナスは不要な気がします

File: configs/sbp_in_domain.yaml (hunk 2 above)
Comment (ciela) [reply configs/sbp_in_domain.yaml:61]: 現状on-policyの方はそうなりますー
ココらへんもon/off-policyでそれぞれ設定ファイル分けたほうが良さそうですね・・・

File: joeynmt/model.py (hunk 3 above)
Comment (ciela) [reply joeynmt/model.py]: 確かに、、、修正しておきました！

**Review by tetsuromorimura — APPROVED** [review]
ありがとうございます！問題なさそうです！
